mgplvm/rdist/rGP.py

```diff
@@ -109,25 +97,27 @@ def prms(self):
 
         #K^(1/2) has sig variance sigma*2^1/4*pi^(-1/4)*ell^(-1/2) if K has sigma^2
         sig_sqr_half = 1 * (2**(1 / 4)) * np.pi**(-1 / 4) * (self.ell**(-1 / 2)
-                                                            )  #(1 x d x 1 x 1)
-        #sig_sqr_half = 1
+                                                            )  #(1 x d x 1)
 
-        #(n_samples x d x m x m)
-        K_half = sig_sqr_half * torch.exp(-self.dts_sq /
-                                          (2 * torch.square(ell_half)))
         # the if and else do the same matmul, but sym_toeplitz takes advantage of structure
         if self.use_fast_toeplitz:
-            mu = sym_toeplitz_matmul(K_half[:, :, :, 0],
-                                     nu[..., None])  #(n_samples x d x m x 1)
+            # (n_samples x d x m)
+            K_half = sig_sqr_half * torch.exp(-self.dts_sq.to(ell_half.device) /
+                                              (2 * torch.square(ell_half)))
+            mu = sym_toeplitz_matmul(K_half, nu[...,
+                                                None])  #(n_samples x d x m x 1)
+
         else:
+            #compute full TxT covariance matrix
+            #(n_samples x 1 x m x m)
+            dts_sq = torch.square(self.ts[..., None] - self.ts[..., None, :])
```

Comment:
the toeplitz file includes ways to generate the full matrix from one column, maybe useful to use these features here?

also I don't fully understand what dts_sq is but can we broadcast / repeat the version that we make by default for toeplitz for the non_toeplitz case rather than regenerating? Or is that slower / not possible?

Reply:
actually if this is generated from the K_half for toeplitz than we don't even need to make the bigger dts.

Reply:
yes in this case I agree that we should just compute a single column of K_half and then construct the full matrix from this if toeplitz = False

Reply:
update - sym_toeplitz is inefficient, we don't see this being used that much so will leave as is.